R/verb-copy-to.R

```diff
@@ -87,6 +87,120 @@ copy_to.src_sql <- function(dest, df, name = deparse(substitute(df)),
   invisible(out)
 }
 
+#' @export
+db_values <- function(con, df) {
+  if (ncol(df) == 0) {
+    abort("`df` needs at least one column.")
+  }
+
+  if (!inherits(df, "data.frame")) {
+    abort("`df` needs to be a data.frame.")
+  }
+
+  tbl(
+    src = con,
+    from = sql_values(con, df),
+    vars = colnames(df)
+  )
+}
+
+sql_values <- function(con, df) {
+  UseMethod("sql_values")
+}
+
+#' @export
+sql_values.DBIConnection <- function(con, df) {
+  sql_values_clause(con, df)
+}
+
+sql_values_clause <- function(con, df, row = FALSE) {
+  # The query consists of two parts:
+  # * An outer select which converts the values to the correct types
+  # * A subquery which provides:
+  #   * named NULLs (needed as e.g. SQLite cannot name `VALUES`)
+  #   * the actual values without names
+
+  sim_data <- rep_named(colnames(df), list(NULL))
+  cols_clause <- escape(sim_data, con = con, parens = FALSE)
+  null_row_clause <- build_sql("SELECT ", cols_clause, " WHERE 0 = 1", con = con)
+
+  empty_df <- nrow(df) == 0L
+  if (empty_df) {
+    subquery <- null_row_clause
+  } else {
+    escaped_values <- purrr::map(df, escape, con = con, collapse = NULL, parens = FALSE)
+    rows <- rlang::exec(paste, !!!escaped_values, sep = ", ")
+    values_clause <- sql(paste0(if (row) "ROW", "(", rows, ")", collapse = ",\n  "))
+
+    union_query <- set_op_query(
+      null_row_clause,
+      build_sql("VALUES\n  ", values_clause, con = con),
+      type = "UNION ALL"
+    )
+
+    subquery <- sql_render(union_query, con = con)
+  }
+
+
+  typed_cols <- purrr::map2_chr(df, colnames(df), ~ select_typed_col(.x, ident(.y), con = con))
+  select_clause <- sql_vector(typed_cols, parens = FALSE, collapse = ", ", con = con)
+
+  build_sql(
+    "SELECT ", select_clause, "\n",
+    "FROM ", sql_subquery(con, subquery, name = "values_table"),
+    con = con
+  )
+}
+
+select_typed_col <- function(type, col, con) {
+  UseMethod("select_typed_col")
+}
+
+#' @export
+select_typed_col.default <- function(type, col, con) {
+  translate_sql(!!col, con = con)
```

Comment:
This feels weird to me — maybe it would be better to create a new translation function that converts from the R type to the database type? And then all the `as.*` function would use that instead of managing their own SQL casts.

Something like:

```R
sql_type <- function(con, type) {
  UseMethod("sql_type")
}

sql_type.DBIConnection <- function(con, type) {
  switch(type, 
    numeric = "NUMERIC",
    double = "NUMERIC",
    integer = "INTEGER",
    character = "TEXT",
    logical = "BOOLEAN",
    Date = "DATE",
    POSIXct = "TIMESTAMP",
    integer64  = "BIGINT",
    stop("Unsupported type")
  )
}
sql_type.odbc <- function(con, type) {
  switch(type, 
    number = "DOUBLE",
    double = "DOUBLE",
    NextMethod()
  )
}
```

Does that make sense?

Reply:
So, `sql_cast()` would basically become

```r
sql_cast <- function(con, r_type) {
  sql_type <- sql_type(con, r_type)
  function(x) {
    sql_expr(cast(!!x %as% !!sql_type))
  }
}
```

Basically, `sql_type()` would be a reimplementation of `DBI:::dbDataType()`, right? E.g. look at the [`RSQLite::dbDataType()`](https://github.com/r-dbi/RSQLite/blob/main/R/dbDataType_SQLiteDriver.R) or [`RPostgres:::get_data_type()`](https://github.com/r-dbi/RPostgres/blob/e9f0366742217b9337979e914215419000970629/R/PqConnection.R).

Reply:
What should `sql_type(con, type)` look like?

1. `type` a type string, e.g. `"Date"` or `"integer"` (like you proposed).
2. `type` a prototype. e.g. `1L` or `Sys.Date()` (like `DBI::dbDataType()`)

The second approach is easier to use in `db_values()`, the first a bit easier for the translations e.g. `as.numeric = sql_cast("Date")` vs `as.numeric = sql_cast(sql_type(vctrs::new_date()))`.

Reply:
Maybe have a mutually exclusive `type` and `ptype` arguments so you could call `sql_type(con, type = "numeric")` or `sql_type(con, ptype = Sys.Date())`?

Reply:
Looking at the different backends I think that it is necessary to use the translation of the `as.*()` functions. For example for MySQL the translation of `as.logical()` is defined as

```r
as.logical = function(x) sql_expr(IF(x, TRUE, FALSE))
```

or MS SQL

```r
as.integer = function(x) {
  sql_expr(try_cast(try_cast(!!x %as% NUMERIC) %as% INT))
}
```

Reply:
So, for MySQL we want this R code

```r
copy_inline(
  simulate_mysql(),
  tibble(
    lgl = TRUE,
    date = Sys.Date()
  )
) %>% 
  show_query()
```

to result in this SQL code

```SQL
SELECT IF(`lgl`, TRUE, FALSE) AS `lgl`, CAST(`date` AS DATE) AS `date`
FROM (
  (
    SELECT NULL AS `lgl`, NULL AS `date`
    WHERE (0 = 1)
  )
  UNION ALL
  (  VALUES (TRUE, '2022-03-18'))
) `values_table`
```

This does not work with `sql_type()` but only by translating `as.logical()` etc. This looks ugly but I don't really see a way around it.

Reply:
Ah good point. The S3 dispatch still feels a bit weird to me, but I don't see a simpler way of doing it, so I think it'd just be worth noting why this design is necessary in a comment.